static_frame/core/index_hierarchy.py

```diff
@@ -662,6 +650,91 @@ def _repeat(i_repeat_tuple: tp.Tuple[int, int]) -> np.ndarray:
                 name=name,
                 )
 
+    @classmethod
+    def from_index_items(cls: tp.Type[IH],
+            items: tp.Iterable[tp.Tuple[tp.Hashable, IndexBase]],
+            *,
+            index_constructor: IndexConstructor = None,
+            name: NameType = None,
+            ) -> IH:
+        '''
+        Given an iterable of pairs of label, :obj:`IndexBase`, produce an :obj:`IndexHierarchy` where the labels are depth 0, the indices are depth 1. While the provided :obj:`IndexBase` can be `Index` or `IndexHierarchy`, across all pairs all depths must be the same.
+
+        Args:
+            items: iterable of pairs of label, :obj:`IndexBase`.
+            index_constructor: Optionally provide index constructor for outermost index.
+        '''
+        items = iter(items)
+        try:
+            label, index = next(items)
+        except StopIteration:
+            return cls._from_empty((), name=name, depth_reference=2)
+
+        if index.depth == 1:
+            return cls._from_index_items_1d(
+                    itertools.chain([(label, index)], items), # type: ignore
+                    index_constructor=index_constructor,
+                    name=name,
+                    )
+
+        assert isinstance(index, IndexHierarchy) # mypy
+
+        depth = index.depth
+        labels: tp.List[tp.Hashable] = [label]
+        repeats: tp.List[int] = [len(index)]
+        existing_index_constructors: tp.List[IndexConstructor] = list(index._index_constructors)
+
+        blocks = [index._blocks]
+        for label, index in items:
+            if index.depth != depth:
+                raise ErrorInitIndex("All indices must have the same shape.")
+
+            labels.append(label)
+            repeats.append(len(index))
+            blocks.append(index._blocks)  # type: ignore
+
+            # If the IndexConstructor differs for any level, downcast to the
+            # default constructor.
+            for i, constructor in enumerate(index._index_constructors):  # type: ignore
+                if constructor != existing_index_constructors[i]:
+                    existing_index_constructors[i] = cls._INDEX_CONSTRUCTOR
+
+
+        outer_level = np.hstack(
+                [
+                    np.repeat(val, repeats=reps)
+                    for val, reps in zip(labels, repeats)
+                ])
```

Comment:
For most common cases this will work fine, but in some cases we will hit NumPy's undesirable type evaluation procedures, like this:
```
>>> np.hstack([np.repeat(v, repeats=r) for v, r in zip([1, 'b'], [3, 4])])
array(['1', '1', '1', 'b', 'b', 'b', 'b'], dtype='<U21')
```
I think there is the no alternative but to use `iterable_to_array_1d()` for the final array creation... or evaluate the types per element first and resolve them with `resolve_dtype`, then fill the value into a pre-allocated array.